### PR TITLE
fix: the fireviewabilitypixels function directly int... in index.js

### PR DESCRIPTION
**Pull request by anupamme**
## Summary
Fix high severity security issue in `libraries/bidViewabilityPixels/index.js`.

## Vulnerability
| Field | Value |
|-------|-------|
| **ID** | V-001 |
| **Severity** | HIGH |
| **Scanner** | multi_agent_ai |
| **Rule** | `V-001` |
| **File** | `libraries/bidViewabilityPixels/index.js:42` |
| **Assessment** | Likely exploitable |

**Description**: The fireViewabilityPixels function directly interpolates URLs from bid.eventtrackers into script tags without validation. Malicious bidders can include JavaScript protocol URLs or data URIs that execute arbitrary code when injected into iframes via insertHtmlIntoIframe.

## Evidence

**Exploitation scenario**: Attacker-controlled bidder returns bid response with eventtrackers containing malicious URLs (javascript: or data: URIs).

**Scanner confirmation**: multi_agent_ai rule `V-001` flagged this pattern.

**Production code**: This file is in the production codebase, not test-only code.

## Threat Model Context

This is a Node.js library - vulnerabilities affect downstream consumers who use this package.

## Changes
- `libraries/bidViewabilityPixels/index.js`

## Behavior Preservation
The change is scoped to 1 file on the vulnerable path; it only tightens handling of untrusted input and leaves valid inputs unaffected.

<details>
<summary>Regression test</summary>

```javascript
const { fireViewabilityPixels } = require('../../libraries/bidViewabilityPixels/index.js');

describe("fireViewabilityPixels must not execute arbitrary JavaScript from untrusted URLs", () => {
  const payloads = [
    // Exact exploit case: JavaScript protocol URL
    "javascript:alert('xss')",
    // Boundary case: Data URI with JavaScript
    "data:text/html;base64,PHNjcmlwdD5hbGVydCgneHNzJyk8L3NjcmlwdD4=",
    // Valid input that should be allowed
    "https://trusted-tracker.com/pixel.js"
  ];

  test.each(payloads)("sanitizes or rejects adversarial input: %s", async (payload) => {
    // Create a mock bid with the payload in eventtrackers
    const mockBid = {
      eventtrackers: [
        {
          url: payload,
          type: 1 // Assuming type 1 represents viewability tracking
        }
      ]
    };

    // Create a mock iframe to capture injected HTML
    let capturedHtml = null;
    const mockInsertHtmlIntoIframe = (html) => {
      capturedHtml = html;
    };

    // Temporarily replace the global insertHtmlIntoIframe function
    const originalInsertHtmlIntoIframe = global.insertHtmlIntoIframe;
    global.insertHtmlIntoIframe = mockInsertHtmlIntoIframe;

    try {
      // Call the production function with adversarial input
      fireViewabilityPixels(mockBid);

      // Security property: No script tags should contain JavaScript protocol or data URIs
      if (capturedHtml) {
        const hasJavaScriptProtocol = capturedHtml.includes('javascript:');
        const hasDataUri = capturedHtml.includes('data:text/html');
        
        // Assert that neither dangerous pattern appears in the generated markup
        expect(hasJavaScriptProtocol).to.be.false;
        expect(hasDataUri).to.be.false;
        
        // Additional safety check: If URL was dangerous, it should either be
        // sanitized (e.g., removed or made inert) or not included at all
        if (payload.startsWith('javascript:') || payload.startsWith('data:text/html')) {
          // The URL should not appear in the script src attribute
          const scriptTagContainsPayload = capturedHtml.includes(payload);
          expect(scriptTagContainsPayload).to.be.false;
        }
      }
    } finally {
      // Restore original function
      global.insertHtmlIntoIframe = originalInsertHtmlIntoIframe;
    }
  });
});
```

</details>

This test guards against regressions — it's useful independent of the code change above.

---
*Automated security fix by [OrbisAI Security](https://orbisappsec.com)*

### CODE REVIEW

File: libraries/bidViewabilityPixels/index.js [hunk 1]
```diff
@@ -43,8 +43,9 @@ export function fireViewabilityPixels(bid) {
   }
 
   img.forEach(triggerPixel);
-  if (js.length > 0) {
-    const markup = js.map(url => `<script async src="${url}"></script>`).join('\n');
+  const safeJs = js.filter(url => /^https?:\/\//i.test(url));
+  if (safeJs.length > 0) {
+    const markup = safeJs.map(url => `<script async src="${url}"></script>`).join('\n');
```
[the remark this answers is not in the document]
Comment (patmmccann) [reply libraries/bidViewabilityPixels/index.js]: @codex if the attacker has control of the bid response and can inject arbitrary js already, what does it matter if they can do this trick

[the remark this answers is not in the document]
Comment (anupamme) [reply libraries/bidViewabilityPixels/index.js]: Pushed a follow-up fix (`1a4fa68ed`) that addresses this: the `https?://` scheme allowlist is kept for defense-in-depth against `javascript:`/`data:` URIs, and URLs are now HTML-escaped with the existing `attributeValue()` helper (`src/utils/xml.js`, already used by `vastXmlBuilder.js`) before being interpolated into the `src="..."` attribute. That closes the markup-breakout you flagged (e.g. `"></script><script>...`) without touching the scheme check.

@patmmccann re: your question — escaping doesn't and isn't meant to stop an accepted `https://` JS tracker from executing; that's the deliberate JS-tracker feature and out of scope here. What it does stop is that URL string breaking out of the attribute to inject arbitrary additional markup/script tags into the iframe document — a strictly worse primitive (attacker-controlled markup injection) than "the accepted tracker URL's own script runs," and one a bidder shouldn't need control over the response format to pull off. Added a regression test (`bidViewabilityPixels_spec.js`) covering the escaping.

**Review by patmmccann — APPROVED** [review]
